### PR TITLE
[WIP] Implementation of DominantHand

### PR DESCRIPTION
**Pull request by ZephireNZ**
SpongePowered/SpongeAPI#1121

This implements the new `HandType` manipulator, with potential for mods to add arbitrary hands later with Forge.

Test Plugin code:

```COMING SOON TO A THEATRE NEAR YOU```

### CODE REVIEW

File: src/main/java/org/spongepowered/common/registry/type/entity/HandTypeRegistryModule.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import org.spongepowered.api.registry.util.RegisterCatalog;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+
+public class HandTypeRegistryModule implements CatalogRegistryModule<HandType> {
+
+    @RegisterCatalog(HandType.class)
+    public final Map<String, HandType> skinPartMap = Maps.newHashMap();
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/registry/type/entity/HandTypeRegistryModule.java]: Looks like you forgot to rename this when you copied it over :3

File: src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.data.value.mutable.Value;
+import org.spongepowered.common.data.manipulator.mutable.entity.SpongeDominantHandData;
+import org.spongepowered.common.data.processor.common.AbstractEntitySingleDataProcessor;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+import org.spongepowered.common.data.value.mutable.SpongeValue;
+
+import java.util.Optional;
+
+public class DominantHandDataProcessor extends AbstractEntitySingleDataProcessor<EntityLiving, HandType, Value<HandType>, DominantHandData, ImmutableDominantHandData> {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java]: `EntityLivingBase`, not `EntityLiving`.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    public DominantHandDataProcessor() {
+        super(EntityLiving.class, Keys.DOMINANT_HAND);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java]: `EntityLivingBase`, not `EntityLiving`.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Override
+    protected boolean set(EntityLiving dataHolder, HandType value) {
+        dataHolder.setActiveHand(Preconditions.checkNotNull((EnumHand) (Object) value));
+        return true;
+    }
+
+    @Override
+    protected Optional<HandType> getVal(EntityLiving dataHolder) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java]: `EntityLivingBase`, not `EntityLiving`.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.Optional;
+
+public class DominantHandDataProcessor extends AbstractEntitySingleDataProcessor<EntityLiving, HandType, Value<HandType>, DominantHandData, ImmutableDominantHandData> {
+
+    public DominantHandDataProcessor() {
+        super(EntityLiving.class, Keys.DOMINANT_HAND);
+    }
+
+    @Override
+    protected boolean set(EntityLiving dataHolder, HandType value) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java]: `EntityLivingBase`, not `EntityLiving`.

File: src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        dataHolder.setActiveHand(Preconditions.checkNotNull((EnumHand) (Object) value));
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java]: This isn't correct, it depends on the type of `Entity`:
- `EntityPlayer`: `net.minecraft.entity.player.EntityPlayer#setPrimaryHand`
- `EntityLiving` (Mob): `net.minecraft.entity.EntityLiving#setLeftHanded`

File: src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Override
+    protected Optional<HandType> getVal(EntityLiving dataHolder) {
+        return Optional.of((HandType) (Object) dataHolder.getPrimaryHand());
+    }
+
+    @Override
+    protected ImmutableValue<HandType> constructImmutableValue(HandType value) {
+        return ImmutableSpongeValue.cachedOf(Keys.DOMINANT_HAND, DEFAULT_HAND, value);
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java]: use `this.key`

File: src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Override
+    protected Value<HandType> constructValue(HandType actualValue) {
+        return new SpongeValue<>(Keys.DOMINANT_HAND,DEFAULT_HAND, actualValue);
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/processor/data/entity/DominantHandDataProcessor.java]: Use `this.key`
space after comma

File: src/main/java/org/spongepowered/common/data/property/store/entity/DominantHandPropertyStore.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package org.spongepowered.common.data.property.store.entity;
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/data/property/store/entity/DominantHandPropertyStore.java]: Need license header

File: src/main/java/org/spongepowered/common/data/property/store/entity/DominantHandPropertyStore.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import org.spongepowered.api.data.type.HandType;
+import org.spongepowered.common.data.property.store.common.AbstractEntityPropertyStore;
+
+import java.util.Optional;
+
+public class DominantHandPropertyStore extends AbstractEntityPropertyStore<DominantHandProperty> {
+
+    @Override
+    protected Optional<DominantHandProperty> getForEntity(Entity entity) {
+        if(!(entity instanceof EntityPlayer)) return Optional.empty();
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/data/property/store/entity/DominantHandPropertyStore.java]: Formatting.

**Pull request closed without merging**